src/main/metabase/examples.ts

```diff
@@ -0,0 +1,152 @@
+import { postgresQuote } from '../../shared';
+import { Field, Table } from './types';
+
+type Display = 'table' | 'bar' | 'row' | 'scalar' | 'map'; // Other types TBD.
+
+/** An example query card. */
+type ExampleQuery = {
+  name: string; // Title of card.
+  sql: string; // SQL query.
+  sizeX: number; // Grid of 18 units wide.
+  sizeY: number; // Height of card in units.
+  display: Display;
+  visualizationSettings: Record<string, unknown>; // To be typed later.
+
+  // There's also row/col properties, but we'll make some rectangle
+  // packing algorithm to arrange cards automatically in a section.
+};
+
+/** A section for a group of examples. */
+type ExamplesSection = {
+  title: string | null; // Markdown text as the section heading.
+  queries: ExampleQuery[]; // Cards in section.
+};
+
+type ExampleInfo = {
+  sql: string;
+  name: string;
+};
+
+function lines(...lines: string[]) {
+  return lines.join('\n');
+}
+
+const numberFieldTypes = ['int2', 'int4', 'int8', 'float4', 'float8', 'numeric'];
+
+function rawGroupBySQL(column: string, table: string, displayName: string): ExampleInfo {
+  return {
+    name: `${displayName} by ${column}`,
+    sql: lines(`SELECT ${postgresQuote(column)}`, `FROM ${postgresQuote(table)}`),
+  };
+}
+
+function countDistinctSQL(column: string, table: string): ExampleInfo {
+  return {
+    name: `Distinct ${column}`,
+    sql: lines(`SELECT count(distinct ${postgresQuote(column)}) as distinct_${column}`, `FROM ${postgresQuote(table)}`),
+  };
+}
+
+function avgSQL(column: string, table: string): ExampleInfo {
+  return {
+    name: `Average ${column}`,
+    sql: lines(`SELECT avg(${postgresQuote(column)}) as avg_${column}`, `FROM ${postgresQuote(table)}`),
+  };
+}
+
+function textGeneralizedSQL(column: string, table: string, displayName: string, averageLength: number): ExampleInfo {
+  const nChars = Math.ceil(averageLength / 4);
+  const stars = '*'.repeat(Math.ceil(averageLength - nChars));
+  const bucket = `substring(${postgresQuote(column)}, 1, ${nChars})`;
+
+  return {
+    name: `${displayName} by ${column}`,
+    sql: lines(`SELECT ${bucket} || ${stars}, count(*)`, `FROM ${postgresQuote(table)}`, `GROUP BY ${bucket}`),
+  };
+}
+
+function yearlyGeneralizedSQL(column: string, table: string, displayName: string): ExampleInfo {
+  const bucket = `extract(year from ${postgresQuote(column)})`;
+
+  return {
+    name: `${displayName} by ${column} year`,
+    sql: lines(`SELECT ${bucket} as ${column}_year, count(*)`, `FROM ${postgresQuote(table)}`, `GROUP BY ${bucket}`),
+  };
+}
+
+function makeExampleInfos(field: Field, table: Table, aidColumns: string[]): ExampleInfo[] {
+  try {
+    if (field.semantic_type === 'type/PK' || field.database_type === 'serial') {
+      // No sensible example for columns being just row IDs.
+      return [];
+    } else if (aidColumns.includes(field.name)) {
```

Comment:
I'd split (in a future revision) all these into their own function, i.e. `aidColumnExamples`, `numericColumnExamples`, `textColumnExamples`, etc.

Also a future idea is to get some columns with very few distinct values and try grouping by 2 columns. If the values are distributed enough we might get non-suppressed results.